akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -354,16 +351,120 @@ final class CoordinatedShutdown private[akka] (
     extends Extension {
-  import CoordinatedShutdown.Reason
-  import CoordinatedShutdown.UnknownReason
+  import CoordinatedShutdown.{ Reason, UnknownReason }
 
   /** INTERNAL API */
   private[akka] val log = Logging(system, getClass)
   private val knownPhases = phases.keySet ++ phases.values.flatMap(_.dependsOn)
 
   /** INTERNAL API */
   private[akka] val orderedPhases = CoordinatedShutdown.topologicalSort(phases)
-  private val tasks = new ConcurrentHashMap[String, Vector[(String, () => Future[Done])]]
+
+  private trait PhaseDefinition {
+    def size: Int
+    def run(recoverEnabled: Boolean)(implicit ec: ExecutionContext): Future[Done]
+  }
+
+  private object tasks {
+    private val registeredPhases = new ConcurrentHashMap[String, StrictPhaseDefinition]()
+    private trait TaskDefinition extends Cancellable {
+      private[tasks] def run(recoverEnabled: Boolean)(implicit ec: ExecutionContext): Future[Done]
+    }
+    private object TaskDefinition {
+      def create(phaseName: String, task: () => Future[Done], name: String): TaskDefinition = new TaskDefinition {
+        // This is a vanilla class instead of a case class to avoid default implementations of .hashCode() and .equals
+        // Different code paths could register the same task under the same name multiple times; in this case we want to
+        // run that task as many times as it was registered (minus the number of times those were cancelled), so they must
+        // be distinct in a Set[TaskDefinition].
+
+        private val p = Promise[Done]()
+
+        override private[tasks] def run(recoverEnabled: Boolean)(implicit ec: ExecutionContext): Future[Done] = {
+          if (log.isDebugEnabled) {
+            log.debug("Performing task [{}] in CoordinatedShutdown phase [{}]", name, phaseName)
+          }
+          if (p.isCompleted) {
+            Future.successful(Done)
+          } else {
+            val f = try {
+              task.apply().recover {
+                case NonFatal(exc) if recoverEnabled =>
+                  log.warning("Task [{}] failed in phase [{}]: {}", name, phaseName, exc.getMessage)
+                  Done
+              }
+            } catch { // in case task.apply() throws
+              case NonFatal(exc) if recoverEnabled =>
+                log.warning(
+                  "Task [{}] in phase [{}] threw an exception before its future could be constructed: {}",
+                  name,
+                  phaseName,
+                  exc.getMessage)
+                Future.successful(Done)
+              case NonFatal(exc) =>
+                Future.failed(exc)
+            }
+            p.completeWith(f)
+            f
+          }
+        }
+
+        override def cancel(): Boolean = {
+          val cancelled = p.trySuccess(Done)
+          // now release all cancelled tasks in phase
+          registeredPhases
+            .merge(phaseName, StrictPhaseDefinition.empty, (previous, incoming) => previous.merge(incoming))
```

Comment:
Hmm, so even if it already was completed or cancelled we trigger an update of the registeredPhases, shouldn't that only happen if we succeeded?

Reply:
The update basically only prunes the cancelled tasks out so they can be garbage-collected, but it's not a huge deal either way.